tsinfer/formats.py

```diff
@@ -1183,22 +1185,23 @@ def add_site(
         return site_id
 
     def finalise(self):
-        if self._mode == self.BUILD_MODE:
-            if self._build_state == self.ADDING_POPULATIONS:
-                raise ValueError("Must add at least one sample individual")
-            elif self._build_state == self.ADDING_SAMPLES:
-                self._individuals_writer.flush()
-                self._samples_writer.flush()
-            elif self._build_state == self.ADDING_SITES:
-                self._sites_writer.flush()
-            if self.num_sites == 0:
-                raise ValueError("Must add at least one site")
-            self._build_state = -1
-            if self.sequence_length == 0:
-                # Need to be careful that sequence_length is JSON serialisable here.
-                self.data.attrs["sequence_length"] = float(self._last_position) + 1
-
-        super(SampleData, self).finalise()
+        try:
+            if self._mode == self.BUILD_MODE:
+                if self._build_state == self.ADDING_POPULATIONS:
+                    raise ValueError("Must add at least one sample individual")
+                elif self._build_state == self.ADDING_SAMPLES:
+                    self._individuals_writer.flush()
+                    self._samples_writer.flush()
+                elif self._build_state == self.ADDING_SITES:
+                    self._sites_writer.flush()
+                if self.num_sites == 0:
+                    raise ValueError("Must add at least one site")
+                self._build_state = -1
+                if self.sequence_length == 0:
+                    # Need to be careful that sequence_length is JSON serialisable here.
+                    self.data.attrs["sequence_length"] = float(self._last_position) + 1
+        finally:
+            super(SampleData, self).finalise()
```

Comment:
I don't think this is the way to do it --- we should only be finalising (marking the file as complete and well-formed) if things succeed.